### ISSUE TITLE
Docs say `const` items are always `'static`, but they aren't.

### ISSUE THREAD

**Issue by theemathas** [rust-lang/rust#159162]
### Location (URL)

https://doc.rust-lang.org/std/keyword.const.html

### Summary

The documentation says:
> The only lifetime allowed in a constant is `'static`, which is the lifetime that encompasses all others in a Rust program.

However, the following program has a non-`'static` `const` (associated) item, contradicting the docs. It compiles without errors.

```rust
struct Thing<'a>(&'a i32);
impl<'a> Thing<'a> {
    const DATA: &'a i64 = &1;
}
```

**Comment by kpreid** [comment]
This documentation sounds to me like it wasn’t written with associated constants in mind at all. Associated constants on traits have power beyond just deduplicating literal values. (History: this documentation appeared in 1.31, and associated constants were stabilized in 1.20.)

It should probably be expanded with associated constants in mind, not just to change the claim about lifetimes.

**Comment by ItsNovaHere** [comment]
Might as well point out that the [reference also says this](https://doc.rust-lang.org/reference/items/constant-items.html#r-items.const.static).

I'm wondering if this should produce a warning or something similar as as far as I understand the only way to reference that constant would be through the type which will always be `'static`.

**Comment by kpreid** [comment]
> I'm wondering if this should produce a warning or something similar as as far as I understand the only way to reference that constant would be through the type which will always be `'static`.

No, it is both possible to obtain a constant with an arbitrary lifetime, and potentially necessary (if the type is contravariant or invariant). Example:

```rust
struct Thing<'a>(fn(&'a i32));
impl<'a> Thing<'a> {
    const DROP: Thing<'a> = Thing(drop);
}
fn example<'a>() -> Thing<'a> {
    Thing::DROP
}
```

This program will not compile if the constant’s type is changed to `Thing<'static>`.
